src/js/live-tracker.js

```diff
@@ -0,0 +1,204 @@
+      }
+
+      this.pastSeekEnd_ = 0;
+      this.lastSeekEnd_ = newSeekEnd;
+      this.trigger('seekableendchange');
+    }
+
+    this.pastSeekEnd_ = this.pastSeekEnd() + 0.03;
+
+    if (this.isBehind() !== this.behindLiveEdge()) {
+      this.behindLiveEdge_ = this.isBehind();
+      this.trigger('liveedgechange');
+    }
+  }
+
+  /**
+   * handle a durationchange event on the player
+   * and start/stop tracking accordingly.
+   */
+  handleDurationchange() {
+    if (this.player_.duration() === Infinity) {
+      this.start();
+    } else {
+      this.stop();
+    }
+  }
+
+  /**
+   * start tracking live playback
+   */
+  start() {
+    if (this.started()) {
+      return;
+    }
+
+    this.trackingInterval_ = this.setInterval(this.trackLive_, 30);
+    this.trackLive_();
+  }
+
+  /**
+   * stop tracking live playback
+   */
+  stop() {
+    if (!this.started()) {
+      return;
+    }
+    this.pastSeekEnd_ = 0;
+    this.lastSeekEnd_ = null;
+    this.behindLiveEdge_ = null;
+
+    this.clearInterval(this.trackingInterval_);
+    this.trackingInterval_ = null;
+  }
+
+  /**
+   * A helper to get the player seekable end
+   * so that we don't have to null check everywhere
+   */
+  seekableEnd() {
+    const seekable = this.player_.seekable();
+
+    if (!seekable || !seekable.length) {
+      return Infinity;
+    }
+
+    return seekable.end(0);
+  }
+
+  /**
+   * A helper to get the player seekable start
+   * so that we don't have to null check everywhere
+   */
+  seekableStart() {
+    const seekable = this.player_.seekable();
+
+    if (!seekable || !seekable.length) {
+      return 0;
+    }
+
+    return seekable.start(0);
+  }
+
+  /**
+   * Get the live time window
+   */
+  liveWindow() {
+    return this.seekableEnd() - this.seekableStart();
+  }
+
+  /**
+   * Determines if the player is live, only checks wether this component
+   * is tracking live playback or not
+   */
+  isLive() {
+    return this.started();
+  }
+
+  /**
+   * Determines if currentTime is at the live edge and won't fall behind
+   * on each seekableendchange
+   */
+  atLiveEdge() {
+    return !this.behindLiveEdge();
+  }
+
+  /**
+   * get what we expect the live current time to be
+   */
+  liveCurrentTime() {
+    return this.pastSeekEnd() + this.seekableEnd();
+  }
+
+  /**
+   * Returns how far past seek end we expect current time to be
+   */
+  pastSeekEnd() {
+    return this.pastSeekEnd_ || 0;
+  }
+
+  /**
+   * If we are currently behind the live edge, aka currentTime will be
+   * behind on a seekableendchange
+   */
+  behindLiveEdge() {
+    return this.behindLiveEdge_;
+  }
+
+  started() {
+    return typeof this.trackingInterval_ === 'number';
+  }
+
+  /**
+   * Seek to the live edge if we are behind the live edge
+   */
+  seekToLiveEdge() {
+    if (this.atLiveEdge()) {
+      return;
+    }
+
+    this.player().pause();
+    this.player().addClass('vjs-waiting');
+    this.one('seekableendchange', () => {
```

Comment:
wouldn't we call `seekToLiveEdge` after a `seekableendchange`?

Reply:
I don't think we would really need a function for that though. You would just wait for `seekableendchange` and then seek to the new `seekable.end` to get to the current live edge. This function just does all that for the user and keeps the UI up to date with what is happening.